csharp/src/Google.Protobuf.Test/GeneratedMessageTest.cs

```diff
@@ -131,8 +131,8 @@ public void RoundTrip_Empty()
             // Without setting any values, there's nothing to write.
             byte[] bytes = message.ToByteArray();
             Assert.AreEqual(0, bytes.Length);
-            TestAllTypes parsed = TestAllTypes.Parser.ParseFrom(bytes);
-            Assert.AreEqual(message, parsed);
+
+            MessageParsingHelpers.AssertRoundtrip(TestAllTypes.Parser, message);

```

Comment:
In this file, please also add codepath that uses CodedInputReader (in ParseFrom) for these tests:
MapWithEmptyEntry
MapWithOnlyValue
MapWithOnlyKey_PrimitiveValue
MapWithOnlyKey_MessageValue
MapIgnoresExtraFieldsWithinEntryMessages
MapFieldOrderIsIrrelevant
MapNonContiguousEntries
DuplicateKeys_LastEntryWins
DiscardUnknownFields_RealDataStillRead
DiscardUnknownFields_AllTypes
TruncatedMessageFieldThrows
ExtraEndGroupThrows